### DISCUSSION TITLE
Help with my GarageDoorOpener - Simple Button Output for Opening and Closing

### DISCUSSION THREAD

**Discussion by Ripable0** [HomeSpan/HomeSpan#649] [Q&A]
Dear Homespan Comiunity!

I have a problem with my Homespan code. I'm relatively new to Arduino and am now looking for help.

I'm in the process of using my garden gate as a Homespan accessory without having to do much rewiring. The goal would be to issue a key command for 1 second via the GarageDoorOpener to switch the potential-free contact on my relay. The button command should trigger pin 23 on the ESP32, which is used to control the relay, both when opening and closing.

Currently the code only issues a persistent command, but I don't understand why... Please help me solve my problem...

here is my code: (I've used the example 13 in this awesome docs!! Props to you!!)

### **Example 13.ino :**

`#include "HomeSpan.h" 
#include "DEV_DoorsWindows.h" 

void setup() {

  Serial.begin(115200);

  homeSpan.begin(Category::Bridges,"HomeSpan Bridge");

  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      
  new SpanAccessory();                                                          
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("Gartentor");
    new DEV_GarageDoor();

}

void loop(){
  
  homeSpan.poll();
  
}
`

### **DEV_DoorsWindows.h :**

`#include <HomeSpan.h>

struct DEV_GarageDoor : Service::GarageDoorOpener {
  SpanCharacteristic *current;
  SpanCharacteristic *target;

  int relayPin = 23;
  bool buttonPressed = false;

  DEV_GarageDoor() : Service::GarageDoorOpener(){
    current = new Characteristic::CurrentDoorState(1);
    target = new Characteristic::TargetDoorState(1);

    pinMode(relayPin, OUTPUT);
    digitalWrite(relayPin, HIGH);

    Serial.print("Configuring Garage Door Opener");
    Serial.print("\n");
  }

  boolean update(){
    if(target->getNewVal()==0){
      LOG1("Opening Garage Door\n");
      current->setVal(2);
      buttonPressed = true;
    } else {
      LOG1("Closing Garage Door\n");
      current->setVal(3);
      buttonPressed = true;
    }
    return(true);
  }

  void loop(){
    if(buttonPressed) {
      triggerRelay();
      buttonPressed = false;
    }

    if(current->getVal()==target->getVal())
      return;

    if(target->timeVal()>5000)
      current->setVal(target->getVal());
  }

  void triggerRelay(){
    digitalWrite(relayPin, LOW);
    delay(1000);
    digitalWrite(relayPin, HIGH);
  }
};
`

**Comment by HomeSpan** [comment]
The code above looks like it should work (though you should avoid delay(1000) in a HomeSpan sketch - see Tutorial 14 for an example of how to emulate a pushbutton).

What is not working?  Is the relayPin staying low instead of just toggling low for 1 second?  Have you tested this with an LED attached to the pin instead of a relay?  This will allow you to easily see exactly what is going on.

**Reply by Ripable0** [reply]
Unfortunately I haven't tried an LED yet... 
The pin stays at LOW instead of changing to LOW. The relay remains switched permanently. I also tried changing the pin (in this case pin 12, as this is often recommended as a relay pin. Still the problem remains...)

I'm going to try an LED today to see exactly what happens...

Thanks for your help!
